CONTRIBUTING.md

```diff
@@ -0,0 +1,140 @@
+```
+kubectl get application -n argocd
+```
+
+## idpbuilder Overview
+
+idpbuilder is made of two phases: CLI and controllers.
+
+### CLI
+
+During this phase: 
+1. This is the phase where command flags are parsed and translated into relevant Go structs' fields. Most notably the [`LocalBuild`](https://github.com/cnoe-io/idpbuilder/blob/main/api/v1alpha1/localbuild_types.go) struct.
+2. Create a Kind cluster, then update the kubeconfig file.
+3. Once the kind cluster is started and relevant fields are populated, Kubernetes controllers are started:
+    *  `LocalbuildReconciler` responsible for bootstrapping the cluster with absolute necessary packages. Creates Custom Resources (CRs) and installs embedded manifests.
+    *  `RepositoryReconciler` responsible for creating and managing Gitea repository and repository contents.
+    *  `CustomPackageReconciler` responsible for managing custom packages.  
+4. They are all managed by a single Kubernetes controller manager.
+5. Once controllers are started, CRs corresponding to these controllers are created. For example for Backstage, it creates a GitRepository CR and ArgoCD Application.
+6. CLI then waits for these CRs to be ready.
+
+### Controllers
+
+During this phase, controllers act on CRs created by the CLI phase.
+
+#### LocalbuildReconciler
+
+`LocalbuildReconciler` bootstraps the cluster using embedded manifests. Embedded manifests are yaml files that are baked into the binary at compile time.
+1. Install Bootstrap apps. They are essential services that are needed for the user experiences we want to enable:
+    * Gitea. This is the in-cluster Git server that hosts Git repositories.
+    * Ingress-nginx. This is necessary to expose services inside the cluster to the users.
+    * ArgoCD. This is used as the packaging mechanism. Its primary purpose is to deploy manifests from gitea repositories. 
+2. Once they are installed, create `GitRepository` CRs for bootstrap apps.
+3. Create ArgoCD applications for the apps. Point them to the Gitea repositories. From here on, ArgoCD manages the bootstrap apps.
+
+Once bootstrap apps are installed, it creates the other embedded applications: Backstage and Crossplane.
+1. Create `GitRepository` CRs for the apps.
+2. Create ArgoCD applications for the apps.
+
+
+#### RepositoryReconciler
+
+`RepositoryReconciler` creates Gitea repositories.
+The content of the repositories can either be sourced from Embedded file system or local file system.
+
+#### CustomPackageReconciler
+
+`CustomPackageReconciler` parses the specified ArgoCD application files. If they specify repository URL with the scheme `cnoe://`,
+it creates `GitRepository` CR with source specified as local, then creates ArgoCD application with the repository URL replaced.
+
+For example, if an ArgoCD application is specified as the following.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+spec:
+  source:
+    repoURL: cnoe://busybox
+```
```

Comment:
This is more like the general README. I dont think this belongs to Contrib. this can be a higher level systems architecture doc as well.

for contributing.md, I mainly restrict it to how to build and how to test, with a reference to a separate file with the above content that describes the architecture of the idpbuilder.